manager/volume.go

```diff
@@ -637,7 +637,18 @@ func (m *VolumeManager) Expand(volumeName string, size int64) (v *longhorn.Volum
 			return nil, fmt.Errorf("cannot get request storage")
 		}
 
-		if pvcSpecValue.Cmp(requestedSize) < 0 {
+		longhornStaticStorageClass, err := m.ds.GetSettingValueExisted(types.SettingNameDefaultLonghornStaticStorageClass)
+		if err != nil {
+			return nil, err
+		}
+
+		isStaticStorageClass := false
+		scName := pvc.Spec.StorageClassName
+		if _, err := m.ds.GetStorageClassRO(*scName); err != nil && *scName == longhornStaticStorageClass {
+			isStaticStorageClass = true
+		}
+
+		if pvcSpecValue.Cmp(requestedSize) < 0 && !isStaticStorageClass {
```

Comment:
I am thinking if Longhorn still needs to round up PVC spec size if a static SC is used.

Reply:
changed.

Reply:
Add condition to check static storage class, since volume size can be mounted and expanded as it is

Reply:
You can refactor this function like the following:
```
	kubernetesStatus := &v.Status.KubernetesStatus
	if kubernetesStatus.PVCName != "" && kubernetesStatus.LastPVCRefAt == "" {
		size, err := m.checkAndExpandPVC(kubernetesStatus.Namespace, kubernetesStatus.PVCName, size)
		if err != nil {
			return nil, err
		}
	}
......


func (m *VolumeManager) checkAndExpandPVC(namespace, pvcName string, size int64) (size int64, err error) {
	pvc, err := m.ds.GetPersistentVolumeClaim(kubernetesStatus.Namespace, kubernetesStatus.PVCName)
	if err != nil {
		return nil, err
	}

	// Backward compatibility
	pvcSCName := *pvc.Spec.StorageClassName
	if pvcSCName == longhornStaticStorageClass {
		if _, err := m.ds.GetStorageClassRO(pvcSCName); err != nil {
			if !apierrors.IsNotFound(err) {
				return nil, err
			}
			// Do nothing for old PVCs created by a static StorageClass
			return size, nil
		}
	}

	requestedSize := resource.MustParse(strconv.FormatInt(size, 10))

	// TODO: Should check for pvc.Spec.Resources.Requests.Storage() here, once upgrade API to v0.18.x.
	pvcSpecValue, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
	if !ok {
		return nil, fmt.Errorf("cannot get request storage")
	}
	if pvcSpecValue.Cmp(requestedSize) < 0 {
......
}
```

Reply:
Let me see how to change it, since part of function return `volume` instead of 'size`

Reply:
refactored.